src/main/xerus/monstercat/api/Playlist.kt

```diff
@@ -0,0 +1,89 @@
+package xerus.monstercat.api
+
+import javafx.beans.property.SimpleBooleanProperty
+import javafx.collections.FXCollections
+import javafx.collections.ObservableList
+import mu.KotlinLogging
+import xerus.ktutil.javafx.properties.SimpleObservable
+import xerus.ktutil.javafx.properties.bindSoft
+import xerus.monstercat.api.response.Track
+import java.util.*
+import kotlin.random.Random
+import kotlin.random.nextInt
+
+object Playlist {
+	val logger = KotlinLogging.logger { }
+	
+	val tracks: ObservableList<Track> = FXCollections.observableArrayList()
+	val history = ArrayDeque<Track>()
+	val currentIndex = SimpleObservable<Int?>(null).apply {
+		bindSoft({
+			tracks.indexOf(Player.activeTrack.value).takeUnless { it == -1 }
+		}, Player.activeTrack, tracks)
+	}
+	
+	val repeat = SimpleBooleanProperty(false)
+	val shuffle = SimpleBooleanProperty(false)
+	
+	operator fun get(index: Int): Track? = tracks.getOrNull(index)
+	
+	/** This property prevents songs that come from the [history] from being added to it again,
+	 * which would effectively create an infinite loop */
+	var lastPolled: Track? = null
+	
+	fun getPrev(): Track? = history.pollLast()?.also {
+		lastPolled = it
+		logger.debug("Polled $it from History")
+	} ?: currentIndex.value?.let { get(it - 1) }
+	
+	fun getNext() = when {
+		shuffle.value -> getNextTrackRandom()
+		repeat.value && (getNextTrack() == null) -> tracks[0]
+		else -> getNextTrack()
+	}
+	
+	fun addNext(track: Track) {
+		tracks.remove(track)
+		tracks.add(currentIndex.value?.let { it + 1 } ?: 0, track)
+	}
+	
+	fun add(track: Track) {
+		tracks.remove(track)
+		tracks.add(track)
+	}
+	
+	fun removeAt(index: Int?) {
+		tracks.removeAt(index ?: tracks.size - 1)
+	}
+	
+	fun clear() {
+		history.clear()
+		tracks.clear()
+	}
+	
+	fun setTracks(playlist: Collection<Track>) {
+		history.clear()
+		tracks.setAll(playlist)
+	}
+	
+	fun getNextTrackRandom(): Track {
+		return if (tracks.size <= 1) tracks[0]
+		else tracks[
+			Random.nextInt(0..tracks.lastIndex).let { if(it >= currentIndex.value!!) it + 1 else it } ]
```

Comment:
that is formatted rather weirdly. I'd probably put the line break before the `.let`. Maybe I also have a utility extension function `.if` which might be helpful in that case, try to have a look.

Reply:
``.if`` doesn't exist apparently.

Reply:
I dumped let and used regular vars. Tell me if it's better for you. b532d228370a5f5d59f733dc4020fe368bec0194

Reply:
I agree. But if you add braces to the else please also add them to the if for visual consistency.

Reply:
972579f78ea1cf946e5e2c9fcedb50288bee1ea6